### PR TITLE
[Magnum] OpengGL 3d library port

### PR DESCRIPTION
**Pull request by sigman78**
Added ports to support Magnum C++ OpenGL engine http://mosra.cz/blog/magnum.php and its direct dependencies

### CODE REVIEW

**Comment by ras0219-msft** [comment]
Thanks for the PR!

I got some compiler errors while testing magnum-plugins:
```
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\buildtrees\magnum-plugins\src\magnum-plugins-645b50647d5164c5ec8f5bc83ba2578f6cfe7d80\src\MagnumPlugins\OpenGexImporter\OpenDdl\OpenDdl.cpp(65): warning C4312: 'reinterpret_cast': conversion from 'Magnum::UnsignedInt' to 'void *' of greater size
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\buildtrees\magnum-plugins\src\magnum-plugins-645b50647d5164c5ec8f5bc83ba2578f6cfe7d80\src\MagnumPlugins\OpenGexImporter\OpenDdl\OpenDdl.cpp(1174): error C2039: 'uninitialized_copy': is not a member of 'std'
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\installed\x64-windows-static\include\MagnumExternal/Optional/optional.hpp(930): note: see declaration of 'std'
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\buildtrees\magnum-plugins\src\magnum-plugins-645b50647d5164c5ec8f5bc83ba2578f6cfe7d80\src\MagnumPlugins\OpenGexImporter\OpenDdl\OpenDdl.cpp(1174): error C3861: 'uninitialized_copy': identifier not found
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\buildtrees\magnum-plugins\src\magnum-plugins-645b50647d5164c5ec8f5bc83ba2578f6cfe7d80\src\MagnumPlugins\OpenGexImporter\OpenDdl\OpenDdl.cpp(1175): error C2039: 'uninitialized_copy': is not a member of 'std'
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\installed\x64-windows-static\include\MagnumExternal/Optional/optional.hpp(930): note: see declaration of 'std'
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\buildtrees\magnum-plugins\src\magnum-plugins-645b50647d5164c5ec8f5bc83ba2578f6cfe7d80\src\MagnumPlugins\OpenGexImporter\OpenDdl\OpenDdl.cpp(1175): error C3861: 'uninitialized_copy': identifier not found
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\buildtrees\magnum-plugins\src\magnum-plugins-645b50647d5164c5ec8f5bc83ba2578f6cfe7d80\src\MagnumPlugins\OpenGexImporter\OpenDdl\OpenDdl.cpp(1176): error C2039: 'uninitialized_copy': is not a member of 'std'
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\installed\x64-windows-static\include\MagnumExternal/Optional/optional.hpp(930): note: see declaration of 'std'
2017-06-10T06:32:38.6195191Z D:\vcpkg-live-ci\_work\1\s\buildtrees\magnum-plugins\src\magnum-plugins-645b50647d5164c5ec8f5bc83ba2578f6cfe7d80\src\MagnumPlugins\OpenGexImporter\OpenDdl\OpenDdl.cpp(1176): error C3861: 'uninitialized_copy': identifier not found
```
Perhaps the author forgot to include `<memory>`?

**Comment by sigman78** [comment]
Great finding! Indeed vs2017 regression confirmed. While PR to the author's repository is pending I will provide a temporary patch. Thanks!

**Comment by ras0219-msft** [comment]
Thanks for the update!

It looks like magnum-plugins is putting its DLLs and LIBs inside nested folders:
```
├───bin
│   └───magnum
│       ├───audioimporters
│       │       AnyAudioImporter.conf
│       │       AnyAudioImporter.dll
│       │       AnyAudioImporter.pdb
│       │       DrWavAudioImporter.conf
│       │       DrWavAudioImporter.dll
│       │       DrWavAudioImporter.pdb
│       │
│       ├───fonts
│       │       StbTrueTypeFont.conf
│       │       StbTrueTypeFont.dll
│       │       StbTrueTypeFont.pdb
│       │
│       ├───imageconverters
│       │       AnyImageConverter.conf
│       │       AnyImageConverter.dll
│       │       AnyImageConverter.pdb
│       │       MiniExrImageConverter.conf
...
├───lib
│   └───magnum
│       ├───audioimporters
│       │       AnyAudioImporter.lib
│       │       DrWavAudioImporter.lib
│       │
│       ├───fonts
│       │       StbTrueTypeFont.lib
│       │
```

I'm not really sure how this library is supposed to be used; plugins often mean `LoadLibrary()` calls at runtime instead of normal linking, but afaict the plugins look like normal libraries. If the plugins are supposed to be LoadLibrary'd then the current arrangement is probably fine (ignore the rest of this post 😛). However if
```cpp
// MyProject/main.cpp
#include <Magnum/AnyAudioImporter/AnyImporter.h>

int main()
{
    Magnum::Audio::AnyImporter foo(...);
}
```
is supposed to work then we need to fix some things. The objective is that
- In CMake, you can do the normal `find_package()`
- In VS (MSBuild), you can paste the above code into an empty project and it will build and link with _no further action_ (assuming you've run `vcpkg integrate install`).

The nested folders are a problem because it prevents our MSBuild integration from working correctly. I think these should be moved up to the respective roots: `lib\` and `bin\`. However, simply moving them after the fact _may_ break the installed `MagnumPluginsConfig.cmake` and `FindMagnumPlugins` because it probably relies on relative positions. *[edit: After spending a bit of time looking at them, they might not be based on relative paths, so moving the dlls might actually be fine]*

This also affects the `magnum` package, although less so:
```
├───lib
│   │   Magnum.lib
│   │   MagnumAudio.lib
│   │   MagnumDebugTools.lib
│   │   MagnumMeshTools.lib
│   │   MagnumPrimitives.lib
│   │   MagnumSceneGraph.lib
│   │   MagnumSdl2Application.lib
│   │   MagnumShaders.lib
│   │   MagnumShapes.lib
│   │   MagnumText.lib
│   │   MagnumTextureTools.lib
│   │
│   └───magnum
│       └───audioimporters
│               WavAudioImporter.lib
```

**Comment by sigman78** [comment]
Thanks, Robert!
It's a valid concern. I've looked it up and made a few adjustments for the msbuild compatibility.

Plugins are loaded via LoadLibrary from the direct location provided by the user. Example projects rely on cmake to produce preprocessed configuration.h files with the correct path to the plugin location. I have weighed it up and went ahead with the similar approach - providing baked paths via defines at the library build time. With that little change, it is possible to use plugins in a copy-paste msbuild projects.

```cpp
// MyProject/main.cpp
#include <Magnum/AnyAudioImporter/AnyImporter.h>

int main()
{
    Magnum::Audio::AnyImporter foo(MAGNUM_PLUGINS_IMPORTER_DIR);
}
```


What do you think about it?

**Comment by ras0219-msft** [comment]
I must admit that since I'm not a direct user of the library, I don't really have the context to be sure what's what here. However, I do have two general feelings:
1. Adding additional macros that people will depend upon in their code seems very suspect; it smells a lot like a "feature" more than "compatibility fix". Patches in Vcpkg are _only_ intended for compatibility, not to add more features. We want upstream to retain full creative control of the project.
2. If upstream intends you to pass in the plugin directory as a string at runtime which will be `LoadLibrary()`'d, the layout seems totally fine as-is to me. As an MSBuild/QMake/Buck/NMake/etc user, you will need to craft the config file yourself -- but that seems intended. Plugins do add considerable complexity to the deployment story, so it's very understandable that they won't fit our "normal" library model.

I definitely see how your config-macro patch could be pretty handy for a lot of non-CMake users and I'd encourage you to submit it to the upstream magnum project! It just feels a bit too much like a feature to accept in Vcpkg.

Therefore, I think the contents of 61dbead6 (just the SDL2 change) are the right approach for Vcpkg right now. If that sounds good to you as well, I'll do a manual merge from that commit and close this PR.

**Comment by mosra** [comment]
Hello, the library author here. Great to see it being integrated into vcpkg :)

Just to clarify some things: 

- The plugins, if built as dynamic, are only ever loaded via `LoadLibrary()`. The `#include <MagnumPlugins/AnyImporter/AnyImporter.h>` snippet will only work from plugin code (if that other plugin depends on `AnyImporter` so its DLL gets loaded in advance) or if the plugins were built as static and linked to your project.
- On Windows, because the executables are relocatable by default, it doesn't make much sense to use the `MAGNUM_PLUGINS_DIR` macros, as they effectively hardcode a path. A better solution is to look for the plugins relative to executable location, using e.g. [Corrade::Utility::executableLocation()](http://doc.magnum.graphics/corrade/namespaceCorrade_1_1Utility_1_1Directory.html#ace3c1d46589390f4fe8d40ae38a4565f). So commonly there's a platform-specific `#ifdef` which uses the macros on Unix-like platforms and executable location on Windows.
- The `<memory>` patch for plugins should be integrated in `master` now, so you can remove it from here.
- I'm now proceeding with enabling MSVC 2017 support for the other projects such as `magnum-integration` and `magnum-extras`. What about putting them here as well?
- You might want to enable also all the additional libraries, plugins and tools for `magnum`, so in the end it would look like this:

        -DWITH_AUDIO=ON
        -DWITH_SDL2APPLICATION=ON
        -DWITH_WINDOWLESSWGLAPPLICATION=ON
        -DWITH_WGLCONTEXT=ON
        -DWITH_OPENGLTESTER=ON
        -DWITH_MAGNUMFONT=ON
        -DWITH_MAGNUMFONTCONVERTER=ON
        -DWITH_OBJIMPORTER=ON
        -DWITH_TGAIMAGECONVERTER=ON
        -DWITH_TGAIMPORTER=ON
        -DWITH_WAVAUDIOIMPORTER=ON
        -DWITH_DISTANCEFIELDCONVERTER=ON
        -DWITH_FONTCONVERTER=ON
        -DWITH_IMAGECONVERTER=ON
        -DWITH_MAGNUMINFO=ON
        -DWITH_AL_INFO=ON

- In `magnum-plugins`, there's [a bunch of additional plugins](http://doc.magnum.graphics/magnum/building-plugins.html) that depend on external `libPNG`, `libJPEG`, etc. libraries that people might find useful next to the `stb_*` stuff. I'm not sure how's the state of their ports here and how to link them together.

Okay, this got a bit longer than expected. Don''t want to put any pressure on anyone, though :)

**Comment by sigman78** [comment]
Thanks for your feedback, Vladimir!

I have updated port with the upstream changes and enabled more features in ```magnum```.  Additional plugins in ```magnum-plugins``` and extra libraries will follow perhaps in the another PR :)

**Comment by ras0219-msft** [comment]
@mosra Thanks for making a great library!

> The plugins, if built as dynamic, are only ever loaded via LoadLibrary(). The `#include <MagnumPlugins/AnyImporter/AnyImporter.h>` snippet will only work from plugin code (if that other plugin depends on AnyImporter so its DLL gets loaded in advance) or if the plugins were built as static and linked to your project.

If static linking works, isn't it also reasonable to dynamic link? For example, in the snippet above, I _believe_ that if linking a static version of the plugin would work, then it would also work to link the dynamic version and deploy the DLL into the same folder as the EXE (which vcpkg can do automatically for you). It seems a bit strange to me to support DLLs via LoadLibrary (implying you have exports set up correctly) and static libs (implying that a standard link works) but not normal DLL linking.

---

Let me step back and briefly explain what Vcpkg does for MSBuild so you can hopefully help us figure out how best to package magnum 😄:
- Within a single triplet, we guarantee only _one_ copy of the library will be available. This could be static libs or DLLs but not both. A project only pulls libraries from one triplet.
- We add `include\` to the include path
- We automatically link all `.lib` files inside the top-level `lib\` folder. MSVC's linker will only add DLL dependencies for symbols you actually use, so if you don't actually use a library you won't pull in any of its code.
- We add `lib\` and `lib\manual-link\` to the lib search path, so you can use `#pragma comment` to pull in libraries that aren't safe to always link.
- As a post-build step, we recursively analyze each application's dependencies with `dumpbin` and copy binaries from `bin\` into the output folder so the application can immediately be run. We _don't_ look in subfolders of `bin\` during this process.

*Note: all the above steps are slightly tweaked for debug since debug and release are ABI incompatible*

So, in the case of using a "normal" library like zlib, you just write 
```
#include <zlib.h>
int main() {
    auto ver = zlibVersion();
}
```
and we'll automatically link against `lib\zlib.lib`. Then, as a post-build action, we'll notice you have a reference to `zlib1.dll` in your PE header and we'll copy `bin\zlib1.dll` to your output folder so you can immediately hit f5 and run your application.

Obviously, calls to `LoadLibrary` mess this up quite a bit because we can't do the aforementioned transitive search since the names don't end up in the PE header 😉.

I think that this implies a few things:
- If we support static linking, we should move the lib files _in that configuration_ up to the `lib\` folder. This will make them act like normal libraries -- you can just `#include <MagnumPlugins/AnyImporter/AnyImporter.h>` and everything will work fine.
- In the dynamic configuration we have choices:
  - If we want to only support plugin deployment, we should probably just completely delete the import libs and headers. However, you mentioned that plugins are allowed to link directly to other plugins, so I think that this doesn't work for all cases.
  - If we want to require explicit linking (unlike the vast majority of other libraries in vcpkg), we could lift the lib files up to `lib\manual-link\`. This means that users will be able to `#include` the plugin headers, but they'll get linker errors by default. If they do link the lib files, they will not have the DLLs at runtime so the application will fail to run until they figure out how to deploy the DLLs.
  - If we want to enable automatic linking, we can lift the lib files to `lib\` and the dll files to `bin\`. This _should_ appear to work exactly like you were static linking, except you happen to get more DLLs in your output folder. You would still be able to manually deploy the plugins if you want to go the `LoadLibrary` route, but they wouldn't be pre-arranged in the current folder structure (I don't know how important that is).

Obviously, I'm quite biased towards the last option (it sounds great to me as a general C++ developer!), but I'm not familiar enough with Magnum to know if that's the right choice.

**Comment by codicodi** [comment]
> there's no way for us to (externally) ensure it goes at the end of the compiler line

Wouldn't adding it to the `INCLUDE` environment variable work like that?

**Comment by mosra** [comment]
Wow, thanks a lot for the detailed explanation! Clears up quite a lot of things that I was unsure about for quite a long time.

More about plugins:

The default, most common and most featureful usage of plugins is via `LoadLibrary()` or `dlopen()` -- on platforms that support it. There are features like plugin aliases which allow the user to say just *"give me some JPEG importer"* and based on what's available in given plugin directory, the plugin manager can load a plugin using the `stb_*` family of libraries, `libJPEG`, `DevIL` or just anything. There are also inter-plugin dependencies, so for example an OBJ importer can depend on a JPEG importer. There's no explicit DLL linking involved in that -- the plugin manager just makes sure that `LoadLibrary()` and symbol import for the JPEG importer is called *before* loading the OBJ importer.

OTOH, static linking is mainly for platforms that don't support dynamic libraries or the support is painful -- which is currently Emscripten, iOS and Android, possibly also UWP (I'm not sure). Another use case of static plugins is getting rid of the additional DLLs -- when one wants to just have one self-contained `*.exe`. 

In both dynamic and static case, the usage is mainly *without* including the plugin header at all to keep majority of the code the same for both cases:

```cpp
#include <Corrade/PluginManager/Manager.h>
#include <Magnum/Trade/AbstractImporter.h>

int main() {
    #ifdef DO_I_HAVE_THE_PLUGINS_STATIC
    CORRADE_PLUGIN_IMPORT(StbImageImporter) // for example
    #endif

    Corrade::PluginManager::Manager<Magnum::Trade::AbstractImporter> manager{"<some-dir>"};
    std::unique_ptr<Magnum::Trade::AbstractImporter> importer = manager.loadAndInstantiate("JpegImporter");

    // importer is *some* plugin that knows JPEGs
}
```

For the static case, there is [an additional macro](http://doc.magnum.graphics/corrade/AbstractManager_8h.html#a4ced3bca898533c1f46abcd6ec64342b) that defines an `extern`  symbol (in this case something like `pluginImporter_AnyImporter()`) and calls it in order to ensure the static library gets linked in. Based on your explanation above, I think this should "just work".

I'm not supporting linking an application to dynamic plugins for these reasons:

-   There's a limitation in CMake that disallows this -- the plugins are created as `MODULE` libraries and CMake bails out with "you can't link to `MODULE` libraries" when someone tries that
-   Because the plugins are organized in a subfolder hierarchy, it's usually not possible for the system linker to find them -- one would need to add all of the subfolders to `PATH` or some other nasty thing.
-   There's no difference from the usage PoV -- compared to `LoadLibrary()` way there will still be an additional DLL in the application directory and `#include`ing the plugin directly will *not* expose any additional APIs.
-   As said above, inter-plugin dependencies are okay -- the plugin manager takes care of finding the dependency plugin DLL and loading it properly at a right time.

The subfolder organization is needed, because the plugin manager can't know which DLLs have which interfaces without loading every one of them and looking inside. This is very similar to what Qt does with plugins -- for example the `imageformats` plugins are also organized in a special subfolder. There are some assumptions in most of the code about folder names in which the plugins are, so shuffling the files around may break something.

- - -

TL;DR: I would propose:

-   For dynamic plugins, support just the `LoadLibrary()` case, but not dynamically linking the application to them. That's consistent with restrictions of other platforms.
-   For static plugins, move the `*.lib` files to `lib/` (providing you link them all implicitly). That shouldn't be a problem at all.

**Comment by ras0219-msft** [comment]
As an aside, for Qt I did a [terrible horrible hardcoded hack](https://github.com/Microsoft/vcpkg/blob/c966c3709e65b53580e66fe9d0cab1d4414b0472/ports/qt5/qtdeploy.ps1) that deploys the plugins "automatically".

I must admit I'm nervous about how changing from `x86-windows` to `x86-windows-static` dramatically changes the deployment model and coding requirements -- the rest of Vcpkg is thoroughly designed to separate the concerns of authoring and deployment. I think this is a pretty minor issue though and we *should* go ahead with packaging exactly as you've suggested.

To be clear, that would mean:
- Static builds (`VCPKG_LIBRARY_LINKAGE` = `static`)
  - Move all the `.lib` files up to `lib\`
- Dynamic builds (`VCPKG_LIBRARY_LINKAGE` = `dynamic`)
  - Delete all the header files and lib files for plugins

**Comment by mosra** [comment]
Great :)

In case someone would want to do similar "hack" for Magnum (doesn't feel like a very horrible hack to me ;)), the mapping is as follows (it's not documented anywhere *yet*, sorry):

- `MagnumAudio.dll`
    - `*AudioImporter.dll` plugins derived from `Magnum::Audio::AbstractImporter`, stored in `audioimporters/`
- `Magnum.dll`
    - `*Importer.dll` plugins derived from `Magnum::Trade::AbstractImporter`, stored in `importers/`
    - `*ImageConverter.dll` plugins derived from `Magnum::Trade::AbstractImageConverter`, stored in `imageconverters/`
- `MagnumText.dll`
    - `*Font.dll` plugins derived from `Magnum::Text::AbstractFont`, stored in `fonts/`
    - `*FontConverter.dll` plugins derived from `Magnum::Text::AbstractFontConverter`, stored in `fontconverters/`

Thinking about the whole plugin loading thing and `*_PLUGIN_DIR` macros, I have a few ideas about how the whole process could be simplified so it finds out the plugin directories on its own, while still giving the user full control to override the paths. Will try to integrate it to the library in the following weeks.

**Comment by ras0219-msft** [comment]
@sigman78 Could you make the recommended changes by @mosra?

In static builds (VCPKG_LIBRARY_LINKAGE = static): Move all the .lib files up to lib\
In dynamic builds (VCPKG_LIBRARY_LINKAGE = dynamic): Delete all the header files and lib files for plugins

You'll probably need to `set(VCPKG_POLICY_DLLS_WITHOUT_LIBS enabled)` in `magnum-plugins`.

**Comment by sigman78** [comment]
@ras0219 Alright, I'll see what I can do!

**Comment by sigman78** [comment]
I think now port works as you guys expect it to.

Default variant (shared) produces plugins as DLL under rel/dbg plugin-specific directory w/o headers or export libs.
Static variant places plugins along with the other lib path and keeps header files.

There is one more subtle thing though: Magnum/Audio part depends on OpenAL port which compiles to the shared library even if the static variant is requested. I made an adaptation to produce a proper static library which seems to work okayish. However, perhaps the reason how the port works is about the LGPL restriction on the static linkage.

**Comment by ras0219-msft** [comment]
Thanks for the changes, @sigman78!
